fastapi/src/ksi_oidc_fastapi/session_manager.py

```diff
@@ -0,0 +1,189 @@
+import time
+import secrets
+from typing import Dict, Optional, Any
+from dataclasses import dataclass, asdict
+from fastapi import Request
+
+from ksi_oidc_common.tokens import Tokens
+from .oidc_client import get_oidc_client
+
+import asyncio
+
+@dataclass
+class SessionData:
+    """Data stored in a user session"""
+    tokens: Optional[Tokens] = None
+    nonce: Optional[str] = None
+    oauth_state: Optional[str] = None
+    created_at: float = None
+    last_accessed: float = None
+    
+    def __post_init__(self):
+        if self.created_at is None:
+            self.created_at = time.time()
+        if self.last_accessed is None:
+            self.last_accessed = time.time()
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Convert to dictionary for serialization if needed"""
+        return asdict(self)
+
+    def is_expired(self, max_age: int = 300) -> bool:
+        """Check if session is expired (default 5 minutes)"""
+        return time.time() - self.last_accessed > max_age
+
+    def is_authenticated(self) -> bool:
+        """Check if user is authenticated (has valid tokens)"""
+        return self.tokens is not None
+
+
+class SessionManager:
+    """In-memory session manager for FastAPI OIDC authentication"""
+                
+    def __init__(self, session_timeout: int = 300):
+        self._sessions: Dict[str, SessionData] = {}
+        self.session_timeout = session_timeout
+        self._cleanup_task = None
+        
+        try:
+            loop = asyncio.get_running_loop()
+            self._cleanup_task = loop.create_task(self.cleanup_sessions())
+        except RuntimeError:
+            pass
+    
+    def start_cleanup(self):
+        """Start the cleanup task if not already running"""
+        if self._cleanup_task is None or self._cleanup_task.done():
+            try:
+                loop = asyncio.get_running_loop()
+                self._cleanup_task = loop.create_task(self.cleanup_sessions())
+            except RuntimeError:
+                pass
+    
+    async def cleanup_sessions(self):
+        """Periodic cleanup of expired sessions"""
+        while True:
+            await asyncio.sleep(5) 
```

Comment:
Running this function this frequently sounds unnecessary and might cause performance issues, because each call to `cleanup_expired_sessions()` requires enumerating all sessions (it has linear time complexity).